alacritty/src/display/mod.rs

```diff
@@ -1529,6 +1550,54 @@ impl<T> DerefMut for Replaceable<T> {
     }
 }
 
+/// The frame timer state.
+#[derive(Default)]
+pub struct FrameTimerInfo {
+    /// The optimal delay for the renderer.
+    swap_delay: Option<Duration>,
+
+    /// The time stamp of the previous contiguos frame.
+    frame_timestamp: Option<Instant>,
+
+    /// The refresh rate we've used to compute the `swap_delay`.
+    refresh_interval: Duration,
+}
+
+impl FrameTimerInfo {
+    /// Compute the delay that we should use to achieve the target frame
+    /// rate.
+    pub fn compute_delay(&mut self, refresh_interval: Duration) -> Duration {
+        // If the refresh rate has changed, reset the timings.
+        if self.refresh_interval != refresh_interval || self.swap_delay.is_none() {
+            self.refresh_interval = refresh_interval;
+            self.frame_timestamp = Some(Instant::now());
+            self.swap_delay = Some(refresh_interval);
+            return refresh_interval;
+        }
+
+        // At this point we know that `swap_delay` has some value.
+        let swap_delay = self.swap_delay.unwrap();
```

Comment:
Why not use this:

```rust
let swap_delay = match self.sway_delay.filter(|_| self.refresh_interval == refresh_interval) {
    Some(swap_delay) => swap_delay,
    None => {
        // Reset stuff
    },
};
```

Maybe a little worse for performance, but should basically be irrelevant.